Validator keys tool (RIPD-1310)

wilsonianb opened the pull request:
This ripple-libpp based validator key management tool is intended to
replace the Python manifest tool in the rippled git repository.
Validator master keys are generated and stored in a JSON file (default:
`$HOME/.ripple/validator-keys.json`).
Ephemeral signing keys and corresponding manifest are generated
together using the key file. Manifest sequences are tracked and
automatically incremented in the key file.

Reviewers: @ximinez @scottschurr

ximinez commented on src/main.cpp:
```diff
@@ -0,0 +1,234 @@
+    {
+        if (*sequence <= keys.sequence)
+            throw std::runtime_error (boost::str (boost::format (
+                "Sequence should exceed current sequence (%u).") %
```
`boost::str` and `boost::format` need `#include <boost/format.hpp>`. Not sure how that's working on the non-Windows builds.

wilsonianb replied:
Added the include.
I'll look into using appveyor

scottschurr commented:
Two notes so far...

I'm wondering if things are a little bit too easy to use?  Shouldn't we encourage them to put the result of `create_master_keys` somewhere other than their hard drive, like, say, an encrypted USB stick?  By picking a default on their hard drive (and using that default in other commands) we make it easy for them to be lazy.  I'd be inclined to force them to enter a path.  Just my 2 cents.  A security-minded person might have a better idea.

That said, I tried the command...
```
./validator-keys --keyfile /Users/ScottsMini/.ripple/validator-keys.json create_signing_keys
```
Where the `--keyfile` argument is the default master key file location.  I got in response...
```
boost::bad_any_cast: failed conversion using boost::any_cast
```
So I suspect the `--keyfile` option isn't working right.

scottschurr commented:
The good news is that I tried `-u` and `--unittest`.  They both worked.  Also `revoke_master_keys` looked like it did good stuff.  And when I invoked `create_signing_keys` after `revoke_master_keys` I got a great error message.  So it looks like a lot of stuff is working well.

wilsonianb commented:
`--keyfile` option should be fixed now.

Regarding key security, maybe the docs can encourage running the tool on an air-gapped machine and transferring generated ephemeral signing keys via USB.

scottschurr commented:
The `--keyfile` option is working way better now.  Good work.

scottschurr commented:
AppVeyor CI does't seem to be working: https://ci.appveyor.com/project/ripple/validator-keys-tool/build/1.0.7.  Something about "The directory does not contain a project or solution file."

wilsonianb commented:
@scottschurr I turned on the project on AppVeyor but haven't added the appveyor.yml file yet. I think I am going to wait until the rippled master branch has 0.40.0, so that I can take advantage of @ximinez 's CMake changes.

scottschurr commented on doc/manifest-tool-guide.md:
````diff
@@ -0,0 +1,92 @@
+`validator-keys` tool to create a new ephemeral key pair:
+
+```
+  $ validator-keys create_signing_keys
````
I've been unhappy with the words associated with building validation manifests for a while.  It predates this work.  There are too many names (e.g., ephemeral whatevers) and many of the names used in descriptions don't match what the user types or sees in the cfg file or JSON.

In an effort to clean that up a bit, I'd like to suggest a couple of renames.

- I'm wondering if we should rename the "validator_keys" command to "validation_keys".  All of the JSON and cfg stuff it produces (e.g., "validation_public_key", "validation_seed", and "validation_manifest") uses "validation" in the name.  It would be great to be consistent here.

- The name "create_master_keys" is a good one, since that's what we're actually doing, and we talk about it that way.  But I'd like to suggest that we rename the generated file from "validator-keys.json" to "validation-master-keys.json".  That way the name of the generated file aligns well with the command used to generate the file.  Also the "validation" part of the file name aligns with the "validation_public_key" inside the file.

- Suppose we rename "create_signing_keys" to "create_validation_manifest", since that's what the command produces.  Or we could rename it to "create_validation_seed" if that makes more sense to you.  Either way the command name corresponds with what the command actually produces.

- The "revoke-master-keys" command has a great name.  I don't think any changes are called for there.

These are just suggestions, not demands.  It would also be great to have @mDuo13 weigh in on this, since he'll be stuck documenting the process.

mDuo13 replied:
I like Scott's thinking on these points. I don't presently know enough about the details to provide more insightful commentary, but from what I can see this looks good.

scottschurr replied:
Thanks for looking @mDuo13.

wilsonianb replied:
I agree that this is a good opportunity to improve the names.

What do you think about this?

A validator has two sets of keys: master keys and validation keys used to sign validations.
We keep the program name as `validator-keys` but rename the `create_signing_keys` command to `create_validation_keys`.
`validator-keys.json` could keep its name or change to `validator-master-keys.json`. The `validation_public_key` field in the file can be changed to `master_public_key`.

scottschurr replied:
I'm happy to defer to the professional, @mduo13, on naming issues like this.

I don't feel strongly about this, but I have an inclination to use either "validator" or "validation" consistently in all the naming and the JSON keys rather than flipping between "validation" and "validator".  I suspect at least some of the JSON keys are already cemented into "validation", so we might as well stick with that.  If we stick consistently with "validation" that's one less thing for a user to think about.  And if a user resorts to grepping then it's just one chunk of text they have to grep for.  I don't think "validation" is any less informative than "validator".  If we're not cemented into "validation" then I just have the inclination that we use "validator" or "validation" consistently, at least in this subset of the world.

The hard thing for us developers to remember is that users don't use this stuff very often.  We have our heads submerged in the details as we code it up.  A user wants to get in and out in a hurry (at least on stuff like this).  Everything we can do to minimize what they have to think about will at least have them cursing us less.  Just my two pesos.

scottschurr commented on src/ValidatorKeys.cpp:
```diff
@@ -0,0 +1,182 @@
+#include <json_reader.h>
+#include <ripple/protocol/HashPrefix.h>
+#include <ripple/protocol/Sign.h>
+#include <ValidatorKeys.h>
```
Although it goes against the general guideline of listing include files alphabetically, I'll suggest that you put `<ValidatorKeys.h>` first in the list.  Listing the header that corresponds to this `.cpp` file first helps to ensure that the header can stand alone and does not have dependencies on other headers that it does not manage for itself.  This is a guideline from "Large Scale C++ Software Design" Section 3.2, Physical Design Rules.

wilsonianb commented on src/main.cpp:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    const std::string DEFAULT_KEY_FILE =
+        getEnvVar ("HOME") + "/.ripple/validator-keys.json";
```
What do you think about current directory as a fallback?
```
    std::string const homeDir = getEnvVar ("HOME");
    std::string const defaultKeyFile =
        (homeDir.empty () ? boost::filesystem::current_path ().string () : homeDir) +
        "/.ripple/validator-keys.json";
```

scottschurr replied:
I actually don't have a strong opinion about good behavior in this error condition.  I guess my first instinct was to error out if we don't get a good value from `"HOME"`.  But using the current directory would also be viable.  Mostly I just wanted to make sure we'd thought through what might happen and have a justifiable reaction.

scottschurr commented:
LGTM.  Are you intending to get AppVeyor running before merging?

wilsonianb commented:
I think I can get AppVeyor working once the rippled submodule is updated in ripple-libpp
https://github.com/ripple/ripple-libpp/pull/7
But I'm also fine adding it later.

I would like to get the naming right in this PR. What do you think about this?
https://github.com/wilsonianb/validator-keys-tool/pull/1#discussion_r92906405

wilsonianb commented:
I took inspiration from the current [manifest tool guide](https://github.com/ripple/rippled/blob/develop/doc/manifest-tool-guide.md) in this last [commit](https://github.com/wilsonianb/validator-keys-tool/pull/1/commits/5ed2803ea6085923388cc9b95131b1f6138e96f0).
>Validators use two types of key pairs: master keys and ephemeral keys.

The tool and key file's names are still `validator-keys` and `validator-keys.json` since they relate to both sets of keys and users are likely to be referred to them from the [*Validator Setup* instructions](https://ripple.com/build/rippled-setup/#validator-setup)

The `create_signing_keys` command is now `create_ephemeral_keys`. "ephemeral" seems to be the least ambiguous term for distinguishing from the master keys (vs "signing" or "validation").

I think a lot of the naming problems have been due to inconsistent usage (largely by me). If this scheme is acceptable, I'll try to make sure it is adhered to.

scottschurr commented:
I think this most recent effort is a step in the right direction.  I followed the steps and everything started well.  I have two comments on `create_master_keys`:

1. I think things might be clearer if the generated file were named "validator-master-keys.json", rather than "validator-keys.json".  It aligns with the command and suggests to the user that the file is important.
2. The "master_secret" in the generated file might want to be named "master_secret_key".  That occurred to me as I read the validator-keys-tool-guide talking about "master keys", but when I looked inside the file I only saw one key.  The rename would make it clear that they actually got two keys.

I got less happy when I did the "create_ephemeral_keys" step.  It's output doesn't say anything about "ephemeral".  I see a `[validation_seed]`, a `validation_public_key`, and a `[validation_manifest]`.  So, if we retain these names in the config file, I think this step should really be named either "create_manifest" or "create_validation_seed"; that's what you're creating.

I'm also becoming less convinced that the term "ephemeral" is doing us any favors.  It's a cool word.  But maybe what we're really driving toward is "expendable" or "replaceable", like a disposable razor.

I'm assuming that we don't want to change the `[validation_seed]` and `[validation_manifest]` in the config file.  If that's correct I'll summarize my suggestions like this:

1. Rename "validator-keys.json" to "validator-master-keys" so the contents are better identified.
1. In the generated file rename `master_secret` to `master_secret_key`.
2. Rename the `create_ephemeral_key` step to `create_validation_manifest`.
3. When we talk about ephemeral keys in the validator-keys-tool-guide mention that the ephemeral data ends up being embedded as part of the `validate_manifest`.  And consider using "replaceable" instead of "ephemeral".

I'm no user experience expert, so these are only suggestions, not demands.  I hope that helps.  Sorry to keep dragging this out.

wilsonianb commented:
@scottschurr I've incorporated suggestions 1 & 2.

What if we rename the `create-ephemeral-keys` command to `rotate-keys` or `rotate-validation-keys`? (If the latter, I would look into replacing most or all "ephemeral"s with "validation".)
If we do that, then I think it might make sense to generate both sets of keys in `create-master-keys` (in which case, should it be renamed?)

What the three commands would do:

1. Generate master and initial ephemeral/validation keys
2. Generate new ephemeral/validation keys
3. Revoke master keys

scottschurr commented:
I tried out the changes you made and I like them.  Everything seems to work well.

Just from curiosity I disabled write permissions on "validator-master-keys.json" and ran `create_ephemeral_keys`.  That did not go as I would hope.  It went ahead and generated the output and then complained, "Cannot open key file...".  I think we'd be better off if it just refused to generate the output if it can't update the sequence value.

For me naming the "ephemeral keys command `rotate-keys` or `rotate-validation-keys` is not quite right.  To my mind we're not really rotating anything, since the old validation keys can never be re-used.

I'm trying to put myself in the shoes of some poor operations soul.  Some other ops person has just left the company and so this ops person got the order to fix up the validator config file.  If we're lucky they've been through a validator update before.  So they're thinking, "Hmmm.  I need to edit the config file."  They open the file and look around a bit.  "Oh yeah, I have to replace this `[validation_manifest]` thingy."  Then they have to locate your utility and run it.  I guess that's why I'm biased toward naming the "ephemeral keys" step something that contains "validation_manifest".  

Perhaps the "ephemeral keys" step could be named `generate_validation_manifest` or `generate_validation_seed_and_manifest`?  That draws a distinction between the first step (which actually _creates_ a file) and this step where we're _generating_ text that need to be pasted into the config file.

Personally I prefer keeping the create master key functionality separated from the generate ephemeral key functionality (rather than having `create_master_key` produce both).  I agree that keeping them separate is less convenient for a first time user.  However the advantage of keeping them separate is that it provides a clear separation of responsibilities.  The last thing we want them to do is run `create_master_key` in order to produce a validation_manifest.  Keeping the commands separate also trains the user in how they generate the validation_manifest; they always use the same steps to do so.

All of this is just discussion.  I'm not hard over on how this interface needs to work.

ximinez commented:
Test coverage is pretty low considering you're starting from a clean slate (57%).  Looking at the [report](https://codecov.io/gh/wilsonianb/validator-keys-tool/tree/a8a9ec4f204c5d324b2c2d2cd9f743cc91e19d9d/src), I have a few suggestions:
* Exclude json_reader.* until it's available in ripple-libpp.
* Exclude the unit test source file.
* Move the main.cpp helper functions somewhere testable, then write tests for them
  * Maybe even exclude main.cpp from the coverage report.

ximinez commented:
Since I keep harping on it: https://github.com/ripple/ripple-libpp/pull/8

wilsonianb commented on doc/validator-keys-tool-guide.md:
````diff
@@ -0,0 +1,94 @@
+```
+
+Add the `[validation_seed]` and `[validation_manifest]` values to this
+validator's config and restart rippled.
````
I'm going with option 1

ximinez replied:
I was a big fan of option 3, but I guesssss option 1 will do.

wilsonianb commented on src/ValidatorKeys.cpp:
```diff
@@ -0,0 +1,178 @@
+
+    auto const masterPublicKey = derivePublicKey(keyType, *masterSecret);
+
+    if (! jKeys["sequence"].isIntegral())
```
What if I replace this with:
```
std::uint32_t sequence;
try {
    sequence = jKeys["sequence"].asUInt();
} catch (std::runtime_error& e) {
    throw std::runtime_error (
    "Key file '" + keyFile.string() +
    "' contains invalid sequence: " +
    jKeys["sequence"].toStyledString());
}

ximinez replied:
Yeah, that looks good.

wilsonianb commented on src/ValidatorKeys.h:
```diff
@@ -0,0 +1,132 @@
+    operator==(ValidatorKeys const& rhs) const
+    {
+        return sequence_ == rhs.sequence_ && keyType_ == rhs.keyType_ &&
+               masterPublicKey_ == rhs.masterPublicKey_;
```
`ripple::SecretKey` doesn't have comparison operators. What do you think about encoding `masterSecret_` to base58 to compare?

ximinez replied:
I think encoding is a heck of a lot of overhead for a comparison that's only used in tests.

I suggest add a comparison operator for `SecretKey` to rippled. It looks like it could be identical to the comparison for `PublicKey`. Update this function once it's merged. (No rush.)

wilsonianb commented:
Latest commits address all of @ximinez's feedback.
I need to review all of the naming discussion and reach a conclusion.

scottschurr commented:
If write permissions are removed on validator-master-keys.json, then `create_ephemeral_keys` still creates usable output and complains at the end.  I suggest that you perform all operations first (including updating validator-master-keys.json) and only produce useable output if all operations are successful.

scottschurr commented on doc/validator-keys-tool-guide.md:
```diff
@@ -0,0 +1,99 @@
+the master public key, and the new ephemeral validation public key, and
+it is signed with both the ephemeral and master secret keys.
+The sequence number should be higher than the previous sequence number (if it
+is not, the manifest will be ignored). The `validator-keys` tool will
```
The tiniest of nits: I suggest changing this sentence from passive to active voice: "The `validator-keys` tool automatically increments the sequence number."

scottschurr commented on doc/validator-keys-tool-guide.md:
```diff
@@ -0,0 +1,99 @@
+
+Add the `[validation_seed]` and `[validation_manifest]` values to this
+validator's config and restart rippled. After rippled has run for at least a
+few minutes, rename the old key file and generate new [master](#master-keys) and
```
I'm concerned about the phrase, "after rippled has run for at least a few minutes".

You should let them know what they are accomplishing by running their validator this way.  They need to run _synchronized_ to the network for at least a few minutes, right?  Is there a way they can tell the the revocation has spread sufficiently across the network?

To me the "at least a few minutes" guidance is not sufficient.  If I'm an ops person just trying to get something over and done with, running rippled for just a few minutes may not even get the server synced.  They may not accomplish the goal.  If I'm a person worried about really revoking the old key then saying "at least a few minutes" does not give me a feeling for how many minutes will be sufficient (and why).  In one sense the more minutes they do this the more minutes their validator is off line.  In the other sense they want to run this way long enough to make sure the key is really revoked on the network.

wilsonianb replied:
@scottschurr Great points. Let me think this over.

ximinez commented on src/ValidatorKeys.cpp:
```diff
@@ -0,0 +1,191 @@
+    try {
+        sequence = jKeys["sequence"].asUInt();
+    }
+    catch (std::runtime_error& e)
```
Variable `e` is unused. It looks like you don't want `e.what()` in the error message, so you can just catch without it: `catch (std::runtime_error&)`.

ximinez commented on src/test/ValidatorKeys_test.cpp:
```diff
@@ -0,0 +1,298 @@
+            o.close();
+        }
+
+        std::string error;
```
`error` is unused.

wilsonianb commented:
Sorry for the radio silence.
I'm going to let this sit until I get manifest renaming (RIPD-1386) changes into rippled.

ximinez commented:
No problem. Would it be worth closing the PR and reopening it when you're ready?

wilsonianb commented:
We're back!
I've squashed and added functionality for tokens and revocations:
https://github.com/ripple/rippled/pull/1975
https://github.com/ripple/rippled/pull/2019

scottschurr commented:
I'm assuming that the `[FOLD]` commit is the only one that needs review?

wilsonianb commented:
@scottschurr Yes. I squashed all of the previous fold commits prior closing/reopening the pr.

scottschurr commented on src/test/ValidatorKeys_test.cpp:
```diff
@@ -136,60 +135,121 @@ class ValidatorKeys_test : public beast::unit_test::suite
             auto const keyType = KeyType::ed25519;
```
It feels like we mostly, but don't quite, support secp256k1 keys.  The key type is stored in the `validator-keys.json`, but there's no command line option to set the key type.  I'm good with that state of affairs, if that's where you want to leave it.  But I'm thinking you could omit the `["key_type"]` field from the `validator-keys.json` and raise fewer questions if someone were to peek inside the file.  Your call.

wilsonianb replied:
Good observation. At one point, I was considering omitting the `"key_type"` field and determining it from the public key (`publicKeyType()`). I'm going to take a closer look at that option.

wilsonianb replied:
One issue with storing `"public_key"` and not `"key_type"` is that you could replace the `"public_key"` in the file with a public key derived from the same secret key using a *different key type*, and the tool would have no way of knowing to complain.

I'm going to leave `["key_type"]` in the file, partly to future-proof and partly because the documentation no longer encourages the user to open/inspect the file.

scottschurr commented on src/test/ValidatorKeysTool_test.cpp:
```diff
@@ -59,26 +59,25 @@ class ValidatorKeysTool_test : public beast::unit_test::suite
 
         using namespace boost::filesystem;
 
         std::string const subdir = "test_key_file";
```
Before this point you could test creating a token when the `validator-keys.json` file is missing.

scottschurr commented on src/test/ValidatorKeysTool_test.cpp:
```diff
@@ -88,34 +87,51 @@ class ValidatorKeysTool_test : public beast::unit_test::suite
+
+        using namespace boost::filesystem;
+
+        std::string const subdir = "test_key_file";
```
Before this point you could test creating a revocation when the `validator-keys.json` file is missing.

scottschurr commented on src/ValidatorKeys.cpp:
```diff
@@ -94,41 +108,40 @@ ValidatorKeys::make_ValidatorKeys (
             "Key file '" + keyFile.string() +
-            "' contains invalid master secret: " +
-            jKeys["master_secret_key"].toStyledString());
+            "' contains invalid secret: " +
```
I'd go ahead and say `...invalid secret **key**: `.

wilsonianb commented on src/ValidatorKeys.cpp:
```diff
@@ -94,41 +108,40 @@ ValidatorKeys::make_ValidatorKeys (
+            throw std::runtime_error ("");
 
-    if (! jKeys["sequence"].isIntegral())
+        tokenSequence = jKeys["token_sequence"].asUInt();
```
That's tested here:
https://github.com/wilsonianb/validator-keys-tool/blob/089e660790c8c67e194fae695ce0378ccb8160d0/src/test/ValidatorKeys_test.cpp#L153

scottschurr commented:
[FOLD] commit looks great :+1:

scottschurr commented:
Second `[FOLD]` looks good to me too. :+1:

ximinez commented on src/ValidatorKeys.cpp:
```diff
@@ -0,0 +1,227 @@
+    , tokenSequence_ (tokenSequence)
+    , revoked_ (revoked)
+{
+    publicKey_ = derivePublicKey(keyType_, secretKey_);
```
```
assert((!revoked_) == (tokenSequence_ != std::numeric_limits<std::uint32_t>::max ()));
if(revoked_)
  tokenSequence_ = std::numeric_limits<std::uint32_t>::max ();
else if (tokenSequence_ == std::numeric_limits<std::uint32_t>::max ())
  revoked_ = true;
```

Basically, my concern is that if those values ever drift out of sync (say a user edits the key file), bad things could happen.

wilsonianb replied:
I actually added the `revoked` field to give users an easy way to recover if they "accidentally" revoke the keys (without actually adding the revocation to rippled.cfg). They can simply set `revoked` to false in their validator keys file, and their previous `token_sequence` wouldn't have been lost.
I also don't want to tell the user that their keys are revoked, if they haven't generated a revocation. So even if they muck up the `token_sequence`, the tool will tell them if it can't generate a token, but it will only indicate that the keys are revoked if `revoked` is true.

ximinez commented on src/test/ValidatorKeys_test.cpp:
```diff
@@ -0,0 +1,361 @@
+
+            for (auto const tokenKeyType : keyTypes)
+            {
+                auto const token = keys.createValidatorToken (tokenKeyType);
```
```
if(!BEAST_EXPECT(token))
  continue;
```
